Version change Mesan

docNord opened the pull request:
Closes #97 

Main contributions:
- [x] Update to API version 2.1 from 1.2 for Mesan
- [x] Fix Mesan unit and integration tests
- [x] Generalize Metobs integration tests
- [x] Update to latest ruff version in pyproject.toml

docNord commented:
https://opendata-download-metanalys.smhi.se/api/category/mesan2g/version/1/geotype/point/lon/16/lat/57/data.json

docNord commented:
```
from smhi import mesan
client = mesan.Mesan()
client.get_point(57,16)
```
Works. Thus, the basic levels of the API are now working with the naïve version update I introduced. I will now go through the failing tests to try to piece together what is wrong and why.

docNord commented:
Two strange errors I'm getting @mgcth : 
1) Metobs integration test is failing even though I haven't touched those files
2) I have made only minor changes, but lint is failing under the headline "format with ruff" for almost all files in the library

docNord commented:
This is what it says on my computer, regarding the ruff problem:
```
>>ruff format --check src tests
15 files already formatted
```

mgcth commented:
> 2\. I have made only minor changes, but lint is failing under the headline "format with ruff" for almost all files in the library

This is probably relevant here.

docNord commented:
Ok, will look into it when I get back from Vasaloppet 😊

docNord commented:
The problem in metobs integration seems to be that dates are mismatching. The way dates for reference are parsed is somewhat wishwash - it seems there is an assumption "corrected-archive" will always measure up until a set date ahead of today, which no longer seems to be the case.
I will ponder a solution to the problem.

docNord commented:
Will use direct webb call to get the data from the corrected archive and use the actual date from the actual data to parse in the test, something like:
```data = pd.read_csv("https://opendata-download-metobs.smhi.se/api/version/latest/parameter/1/station/188800/period/corrected-archive/data.csv",skiprows=9, usecols=["Datum","Tid (UTC)"],sep=";")```

docNord commented:
My updated code works for test1 however the problem remains even there. This is due to the fact that the latest date of the actual data in this file is indeed different from what i get from the ifk-smhi code.

docNord commented:
...and that is because I'm not reading the header but the latest data point.

mgcth commented on tests/integration/test_integration_metobs.py:
```diff
@@ -3,16 +3,45 @@
         if i == 1:
-            METOBS_INTEGRATION[i]["Tidsperiod (t.o.m)"] = (
-                datetime.date.today().strftime("%Y-%m") + "-01 07:20:09"
+            METOBS_INTEGRATION[i]["Tidsperiod (t.o.m)"] = directdateread(
```
Is this generalisation (and extra logic) needed?

It looks to me like all failing tests are related to station `192840` which has changed `Tidsperiod`, for some reason? We could probably simplify the tests by removing many of these flaky asserts.

mgcth replied:
I mean, this was a mistake from my side when creating the test. The code still works fine even if the test fail because of the date. That is, our tests should fail if something fails (read format changes or something we can't parse) and not if a date changes.

To simplify this, we could simply remove some asserts.

docNord replied:
Agree. Will simplify and remove some stuff tomorrow. Hanging with wife now 🥰

mgcth reviewed (commented):
Nice work! See some comments below.

docNord commented:
Alright @mgcth, I've simplified and fixed also the ```Metobs``` integration tests now, but I am unable to locally reproduce, or even find a meaningful answer as to what ruff finds appalling with the formatting of 14 out of 15 files, so I'm not getting any further with solving the lint issues. Care to weigh in?

mgcth commented on tests/integration/test_integration_metobs.py:
```diff
@@ -113,11 +97,6 @@ def test_integration_metobs_first(
             header_0
         """
-        if raw_header_0:
-            raw_header_0 = raw_header_0.replace(
```
If raw header isn't used any more, it could be deleted from the parametrisation above.

docNord replied:
Fixed in https://github.com/Ingenjorsarbete-For-Klimatet/ifk-smhi/pull/98/commits/bf5f6d2c6a51ae6ce0e3afdc9ddeb66846a73cd8

mgcth commented on tests/integration/test_integration_metobs.py:
```diff
@@ -134,6 +113,8 @@ def test_integration_metobs_first(
         assert client.data.title == data_title
         if table:
             assert data.iloc[table_locr, table_locc] == table
+            data_header.pop("Tidsperiod (t.o.m)")
+            header_0.pop("Tidsperiod (t.o.m)")
```
Is this due to different datetimes in the stored dataset?

docNord replied:
The headers contain different information to the stored one used for reference. 

Your initial solution was to take todays date and move back a set amount, but that failed when the set amount for whatever reason no longer matched the data. I tried various ways of reading the current date directly, but in the end settled for simply removing the information about the "to"-date from the comparison. The integration test hence checks that everything in the header __but__ the latest date matches, which I think is good enough.

mgcth commented on tests/integration/test_integration_metobs.py:
```diff
@@ -289,7 +265,12 @@ def test_integration_metobs_second(
+            try:
+                header_0.pop("Tidsperiod (t.o.m)")
+            except KeyError:
+                pass
```
Perhaps we should at least log the error?

docNord replied:
Any suggestions as to where and how?

mgcth replied:
```
import logging

logger = logging.getLogger(__name__)



logger.info("...")
```

docNord replied:
Ok. I am somewhat familiar with logging, but for whom are we storing these logs? This is a somewhat ugly solution to the problem that the variable is somehow persistent between calls. I can only pop the end date once, essentially. So rather than importing logging and going through all of the hassle of setting up a log file or whatever, perhaps what you are really suggesting is a neater formulation, something along the terms of
```
if exist(header0["end date"]):
    header0["end date"].pop()
```
?

docNord replied:
Changed as per my suggestion in https://github.com/Ingenjorsarbete-For-Klimatet/ifk-smhi/pull/98/commits/2cd1d5367480da60da64ced32cee3947529fbd9d

mgcth reviewed (commented):
Nice improvements! Just a few more things to remove, I think.

mgcth commented:
> Alright @mgcth, I've simplified and fixed also the `Metobs` integration tests now, but I am unable to locally reproduce, or even find a meaningful answer as to what ruff finds appalling with the formatting of 14 out of 15 files, so I'm not getting any further with solving the lint issues. Care to weigh in?

Locally, you probably have a different `ruff` version.  I had `0.1.*` which I expect because we use compatible releases (https://peps.python.org/pep-0440/#compatible-release) in our pyproject.toml file. If you look at the build for lint in this PR you will see it installs `0.3.2` (https://github.com/Ingenjorsarbete-For-Klimatet/ifk-smhi/actions/runs/8219939531/job/22478471060#step:4:39). I don't know why. It does so locally for me when I recreate the env, too.

Looks like the lint run in main respected the compatible release version: https://github.com/Ingenjorsarbete-For-Klimatet/ifk-smhi/actions/runs/7680658453/job/20932961862#step:4:41

docNord commented:
I changed my local ruff to 0.1.* to match our pyproject. Will change again to match what is running then. Could you have a look at this and try to have github actions run with the defined versions of our pyproject file @mgcth ?

mgcth commented:
> I changed my local ruff to 0.1.* to match our pyproject. Will change again to match what is running then. Could you have a look at this and try to have github actions run with the defined versions of our pyproject file @mgcth ?

As you can tell from my comment above, I don't know why this is happening. `~=` isn't respected any more.

mgcth commented on tests/integration/test_integration_metobs.py:
```diff
@@ -269,11 +241,6 @@ def test_integration_metobs_second(
             raw_header_0
             header_0
         """
-        if raw_header_0:
```
There is a raw_header still present here as a parameter. Why are we checking it in this test if not in the previous?

docNord replied:
I honestly don't know. I have just fixed the problems that have turned up.

mgcth replied:
We should probably remove it here too, then.

mgcth reviewed (approved):
Great job!
